### PR TITLE
Update OpenConnectionFromInterfaces:

### PR DESCRIPTION
**Pull request by jacobweinstock**
Currently, if an implementation that opens a stateful session in its `Opener` implementation fails when open is called and then has a method called, for example, `PowerStateGet`, that needs to use the session info, we will get a panic.

This code updates the `OpenConnectionFromInterfaces` to remove any provider implementations that fail to open successfully. We do this and then returning a new `registrar.Drivers` from the Open method in `client.go`.

This changes the existing `client.go` [`Open`](https://github.com/bmc-toolbox/bmclib/blob/80581fa111f565f69b223b0d5800951638130d5f/client.go#L74) method signature. Returning the `registrar.Drivers` instead of updating in place I think is important. We don't want to hide the "magic" because this can be confusing why providers are being removed. It also allows the end-user to have flexibility in deciding if/when to use this newly returned `registrar.Drivers`. 

New usage:

```go
cl := NewClient(host, port, user, pass)
var err error
cl.Registry.Drivers, err = cl.Open(ctx)
if err != nil {
	panic(err)
}
```

### CODE REVIEW

**Comment by jacobweinstock** [comment]
Hey @joelrebel, this one is ready for you whenever you have a moment. Thanks!

**Review by joelrebel — CHANGES_REQUESTED** [review]

File: bmc/connection_test.go [hunk 1]
```diff
@@ -78,21 +52,37 @@ func TestOpenConnectionFromInterfaces(t *testing.T) {
...
 				badImplementation := struct{}{}
-				generic = []interface{}{&badImplementation}
+				generic = append(generic, &badImplementation)
 			} else {
-				testImplementation := connTester{}
-				generic = []interface{}{&testImplementation}
+				testImplementation := &connTester{MakeErrorOut: tc.makeErrorOut}
+				generic = append(generic, testImplementation)
+				ok := generic[0].(Opener)
+				t.Log(ok)
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply bmc/connection_test.go:60]: ah yes, thanks. will update

File: bmc/power.go [hunk 2]
```diff
@@ -50,7 +53,7 @@ Loop:
 			break Loop
 		default:
 			if elem.powerSetter != nil {
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply bmc/power.go:55]: I like it, will update. thanks!

[the remark this answers is not in the document]
Comment (jacobweinstock) [reply bmc/power.go]: updated

File: bmc/connection_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply bmc/connection_test.go]: updated

**Review by joelrebel — APPROVED** [review]

**Comment by joelrebel** [comment]
LGTM, thanks for this!

**Pull request merged**